### PR TITLE
delete everywhere sending (WEBAPP-3081/3082/3087)

### PR DESCRIPTION
**Pull request by herrmannplatz**

### CODE REVIEW

File: test/unit_tests/conversation/ConversationRepositorySpec.coffee [hunk 1]
```diff
@@ -447,23 +454,6 @@ describe 'z.conversation.ConversationRepository', ->
...
-        id: message_et.id
-        type: z.event.Backend.CONVERSATION.MESSAGE_DELETE
-        conversation: conversation_et.id
-
-      conversation_repository.message_deleted event
-
-      expect(conversation_service.delete_message_from_db).toHaveBeenCalledWith conversation_et.id, message_et.id
-      expect(conversation_et.get_message_by_id message_et.id).toBeUndefined()
-
   #@formatter:on
```
Comment (herrmannplatz) [inline test/unit_tests/conversation/ConversationRepositorySpec.coffee:457]: covered by the other tests

File: app/script/view_model/ModalsViewModel.coffee [hunk 2]
```diff
@@ -32,6 +32,7 @@ z.ViewModel.ModalType =
   CONNECTED_DEVICE: '.modal-connected-device'
   CONTACTS: '.modal-contacts'
   DELETE_MESSAGE: '.modal-delete-message'
+  DELETE_EVERYONE_MESSAGE: '.modal-delete-everyone-message'
```
Comment (bennycode) [inline app/script/view_model/ModalsViewModel.coffee]: I don't like that everyone can be deleted. :cry:

File: app/script/view_model/MessageListViewModel.coffee [hunk 3]
```diff
@@ -471,6 +471,9 @@ class z.ViewModel.MessageListViewModel
     if message_et.is_deletable()
       entries.push label: z.string.conversation_context_menu_delete, action: 'delete'
 
+    if message_et.user().is_me
+      entries.push label: z.string.conversation_context_menu_delete_everyone, action: 'delete-everyone'
```
Comment (bennycode) [inline app/script/view_model/MessageListViewModel.coffee]: We agreed on using `{ }` for this case.

File: app/script/conversation/ConversationService.coffee [hunk 4]
```diff
@@ -279,6 +279,26 @@ class z.conversation.ConversationService
         reject error
 
   ###
+  Load conversation event.
+
+  @param conversation_id [String] ID of conversation
+  @param message_id [String]
+  ###
+  load_event_from_db: (conversation_id, message_id) ->
```
Comment (bennycode) [inline app/script/conversation/ConversationService.coffee:287]: We need a test case for this.

File: test/unit_tests/conversation/ConversationRepositorySpec.coffee [hunk 5]
```diff
@@ -291,44 +291,51 @@ describe 'z.conversation.ConversationRepository', ->
...
-      expect(conversation_repository._delete_message).toHaveBeenCalled()
-      expect(conversation_repository._add_delete_message).not.toHaveBeenCalled()
+      .then ->
+        expect(conversation_repository._delete_message).toHaveBeenCalledWith conversation_et, message_to_delete_et.id
+        expect(conversation_repository._add_delete_message).not.toHaveBeenCalled()
+        done()
+      .catch done.fail
 
-    it 'delete message and add delete message if user is not self', ->
+    it 'delete message and add delete message if user is not self', (done) ->
```
Comment (bennycode) [inline test/unit_tests/conversation/ConversationRepositorySpec.coffee]: `it 'delete` -> `it 'deletes` to make it a correct sentence

File: test/unit_tests/conversation/ConversationRepositorySpec.coffee [hunk 6]
```diff
@@ -291,44 +291,51 @@ describe 'z.conversation.ConversationRepository', ->
...
+      message_to_delete_et = new z.entity.PingMessage()
+      message_to_delete_et.id = z.util.create_random_uuid()
+      conversation_et.add_message message_to_delete_et
 
+      spyOn(conversation_repository, 'get_message_from_db').and.returnValue Promise.resolve message_to_delete_et
       spyOn conversation_repository, '_delete_message'
       spyOn conversation_repository, '_add_delete_message'
 
-    it 'delete message if user is self', ->
+    it 'delete message if user is self', (done) ->
```
Comment (bennycode) [inline test/unit_tests/conversation/ConversationRepositorySpec.coffee]: `it 'delete` -> `it 'deletes` to make it a correct sentence

File: app/script/conversation/ConversationService.coffee (hunk 4 above)
Comment (herrmannplatz) [inline app/script/conversation/ConversationService.coffee:287]: done

File: app/script/localization/strings-de.coffee [hunk 7]
```diff
@@ -454,7 +461,7 @@ z.string.de.url_support_history = 'https://support.wire.com/hc/de/articles/20783
 z.string.de.url_support_mic_access_denied = 'https://support.wire.com/hc/de/articles/202590081'
 z.string.de.url_support_mic_not_found = 'https://support.wire.com/hc/de/articles/202970662'
 z.string.de.url_support_screen_access_denied = 'https://support.wire.com/hc/de/articles/202935412'
-z.string.de.url_support_screen_whitelist = 'https://support.wire.com/hc/de/articles/209423889'
+z.string.de.url_support_screen_whitelist = 'https://support.wire.com/hc/en-us/articles/209423889'
```
Comment (gregor) [inline app/script/localization/strings-de.coffee]: This should not be a US article

File: app/script/conversation/ConversationRepository.coffee [hunk 8]
```diff
@@ -99,6 +99,7 @@ class z.conversation.ConversationRepository
     amplify.subscribe z.event.WebApp.SELF.CLIENT_ADD, @on_self_client_add
     amplify.subscribe z.event.WebApp.USER.UNBLOCKED, @unblocked_user
     amplify.subscribe z.event.WebApp.CONVERSATION.MESSAGE.DELETE, @delete_message
```
Comment (gregor) [inline app/script/conversation/ConversationRepository.coffee]: Should we rename this to `z.event.WebApp.CONVERSATION.MESSAGE.DELETE_SELF` for additional clarity down the road?

File: app/script/conversation/ConversationService.coffee (hunk 4 above)
Comment (bennycode) [inline app/script/conversation/ConversationService.coffee:287]: :tada:

File: app/script/conversation/ConversationRepository.coffee [hunk 9]
```diff
@@ -1242,13 +1255,29 @@ class z.conversation.ConversationRepository
...
 
-  @param conversation_et [z.entity.Conversation] Conversation to post the file
-  @param message_et [z.entity.Message] Message object
+  @param conversation_et [z.entity.Conversation]
+  @param message_et [z.entity.Message]
   ###
-  delete_message: (message_et) =>
-    conversation_et = @active_conversation()
+  delete_message_everyone: (conversation_et, message_et) =>
+    if message_et?
```
Comment (gregor) [inline app/script/conversation/ConversationRepository.coffee]: We are calling this via amplify. Safeguard against an undefined `conversation_et` as well?

`if conversation_et and message_et`

File: app/script/conversation/ConversationRepository.coffee [hunk 10]
```diff
@@ -1242,13 +1255,29 @@ class z.conversation.ConversationRepository
...
+        @logger.log "Failed to send delete message for everyone with id '#{message_id}' for conversation '#{conversation_et.id}'", error
+
+  ###
+  Delete message on your own clients.
+
+  @param conversation_et [z.entity.Conversation]
+  @param message_et [z.entity.Message]
+  ###
+  delete_message: (conversation_et, message_et) =>
     if message_et?
```
Comment (gregor) [inline app/script/conversation/ConversationRepository.coffee]: See above

File: app/script/conversation/ConversationRepository.coffee (hunk 8 above)
Comment (bennycode) [inline app/script/conversation/ConversationRepository.coffee]: I would be afraid of calling `DELETE_SELF`... :fearful:

**Pull request merged**